compiler/rustc_middle/src/ty/consts.rs

```diff
@@ -18,18 +20,28 @@ pub use kind::*;
 pub use valtree::*;
 
 /// Typed constant value.
-#[derive(Copy, Clone, Debug, Hash, TyEncodable, TyDecodable, Eq, PartialEq, Ord, PartialOrd)]
+#[derive(Copy, Clone, Debug, TyEncodable, TyDecodable, PartialEq, Eq, Ord, PartialOrd)]
 #[derive(HashStable)]
 pub struct Const<'tcx> {
-    pub ty: Ty<'tcx>,
+    pub(super) ty: Ty<'tcx>,
 
-    pub val: ConstKind<'tcx>,
+    pub(super) val: ConstKind<'tcx>,
 }
 
 #[cfg(all(target_arch = "x86_64", target_pointer_width = "64"))]
 static_assert_size!(Const<'_>, 48);
 
 impl<'tcx> Const<'tcx> {
+    #[inline]
+    pub fn ty(&self) -> Ty<'tcx> {
+        self.ty
+    }
+
+    #[inline]
+    pub fn val(&self) -> ConstKind<'tcx> {
+        self.val
+    }
```

Comment:
Doesn't this cause unecessary Copies?

Reply:
I agree. I will change this to reference and rerun the perf to see if this is the culprit.